src/core/videosource.cpp

```diff
@@ -536,6 +536,21 @@ bool FFMS_VideoSource::DecodePacket(AVPacket *Packet) {
         std::swap(DecodeFrame, LastDecodedFrame);
         if (!(Packet->flags & AV_PKT_FLAG_DISCARD))
             DelayCounter++;
+    } else if (CodecContext->codec_id == AV_CODEC_ID_VP9 && Ret != AVERROR(EAGAIN) && Ret != AVERROR_EOF) {
+        // The internal VP9 BSF does not follow the push/pull documentation properly.
+        // It should keep a queue of data interanlly so further calls can return EAGAIN,
+        // but with VP9, it frees and discards that data when the next packet is sent.
+        // We need to call it a maximum of one extra time, to account for super frames
+        // that have both a visible and invisible frame in one packet. If we don't,
+        // the API happily and silently creates corrupt data output. There are not even
+        // any warnings output by the decoder. Happy days.
```

Comment:
Have you filed a bug report?

Reply:
I still cannot discern what the actual intention of the API is (if this *should* work or not). Is it a bug? Will I get told off for not calling `avodec_receive_frame` until I hit `AVERROR(EAGAIN)` after every packet? I'll file a bug with a small C program to reproduce if it is worth the time.

... But mostly I don't want to deal with Carl and his lack of understanding of anything related to an API and not a CLI tool.

Reply:
I'm not sure what exactly you're trying to do here (not familiar with the code), but I think the doxygen on the decode API is pretty clear. Anything that doesn't match with it is a bug. I know mediacodec had issues (later fixed AFAIK), so these things happen.

How is this new condition even hit, though? It checks for `Ret != 0` above, so in the else branch `Ret == 0` should always hold true.

Yeah, I wasn't seriously suggesting to report it to the ffmpeg _bug tracker_.

Reply:
@wm4 here is a repro program: https://gist.github.com/dwbuiten/595e1e0ea494ec702afb47973b891ff5

And sample to match: https://www.dropbox.com/s/my8yb9k25wg7qae/test.webm?dl=0

Reply:
Nice, I can take a look later (possibly as late as the weekend).

Reply:
See my comment in that gist, you should call avcodec_receive_frame() in a loop until it returns EAGAIN before sending a new packet, as stated in the doxy. That solves the issue for the example.

VP9 because of the split bsf is apparently one of the few (or the only) decoder that returns more than one frame per packet sent.

Reply:
@jamrial 

> You need to drain all the available frames until you get EAGAIN before sending another packet, as the doxy in avcodec.h states.

It does not state this. It, in fact, explicitly says calling it like that is *not required*:

> Using the API as outlined above is highly recommended. But it is also possible to call functions outside of this rigid schema. For example, you can call avcodec_send_packet() repeatedly without calling avcodec_receive_frame(). In this case, avcodec_send_packet() will succeed until the codec's internal buffer has been filled up (which is typically of size 1 per output frame, after initial input), and then reject input with AVERROR(EAGAIN). Once it starts rejecting input, you have no choice but to read at least some output.

Also, the API's author himself (@wm4) outright told me I should be able to do it this way (and it should keep the exising packet data internally.

So, either the doxy or the code needs to be fixed in FFmpeg.

Reply:
> VP9 because of the split bsf is apparently one of the few (or the only) decoder that returns more than one frame per packet sent.

It doesn't it return *one* frame, per packet, since one is invisible. It returns EAGAIN the second time, but still requires it's called.

Reply:
> It does not state this

I see the following

> - Receive output in a loop. Periodically call one of the avcodec_receive_*()
>   functions and process their output:
>   - For decoding, call avcodec_receive_frame(). On success, it will return
>     an AVFrame containing uncompressed audio or video data.
>   - For encoding, call avcodec_receive_packet(). On success, it will return
>   an AVPacket with a compressed frame.
>
>   Repeat this call until it returns AVERROR(EAGAIN) or an error. The
>   AVERROR(EAGAIN) return value means that new input data is required to
>   return new output. In this case, continue with sending input. For each
>   input frame/packet, the codec will typically return 1 output frame/packet,
>   but it can also be 0 or more than 1.

So it can return more than 1 frame per packet.

> It doesn't it return one frame, per packet, since one is invisible. It returns EAGAIN the second time, but still requires it's called.

For the sample you provided it returns 0 two times (and if you check the actual frame data returned, it's different in both), then EAGAIN the third time, for at least five different packets within the 500 iterations of the loop you wrote.
Your patch here "fixes" it by calling receive_frame a second time because that's enough to get the second frame (next call you never do would be EAGAIN). If for some sample the vp9 split bsf were to return a third frame or more, this fix wouldn't work.

> Also, the API's author himself (@wm4) outright told me I should be able to do it this way (and it should keep the exising packet data internally.

But it does. In your repro case you're not checking the return value of avcodec_send_packet. If you do, you'll notice it will return EAGAIN in some cases because the decoder/bsf were not done returning the remaining frames from the previous packet and hadn't yet gotten around to fetch the packet you had previously pushed.

You do:

send_packet 1 -> ret 0;
receive_frame -> packet 1 fetched, ret 0;
send_packet 2 -> ret 0;
receive_frame -> no packet fetched as data from packet 1 is still available and thus used instead, ret 0;
send_packet 3 -> buffer already has packet 2 that was never fetched, so new data is rejected, ret EAGAIN;
receive_packet -> packet 2 fetched, ret 0 (lets assume this is only one frame);
send_packet 4 -> ret 0;
receive_frame -> packet 4 fetched, decoding is ruined because packet 3 was lost.

I admit i may be making a wrong analysis (wm4 is after all the author of this API), but what i know is that ffmpeg and ffplay all handle this sample right by calling receive_frame repeatedly until it returns EAGAIN, and making your repro program do the same also fixes the issue for it.

Reply:
> I see the following

Yes and *directly* below it, it says it isn't explicitly required:

> Using the API as outlined above is highly recommended. But it is also possible to call functions outside of this rigid schema.

Misleading / poor doc at best if it is required.

> For the sample you provided it returns 0 two times (and if you check the actual frame data returned, it's different in both), then EAGAIN the third time, for at least five different packets within the 500 iterations of the loop you wrote.

How exactly can one VP9 superframe contain two visible frames? I thought it was always one invisible and one visible *at most*, per VP9 packet. And this seems to be the case when threads==1.

This behavior changes with threads. With threads == 1, I don't get the corrupt-looking output and i get what seems to be correct number of frames out, from the same code. Is this intended? The sample code provided is hardcoded to 16 threads to show this. Is it possible this is related to thread delay?

> But it does. In your repro case you're not checking the return value of avcodec_send_packet. If you do, you'll notice it will return EAGAIN in some cases because the decoder/bsf were not done returning the remaining frames from the previous packet and hadn't yet gotten around to fetch the packet you had previously pushed.

You say it does, but then describe how it doesn't (since it ditches any data from frames it hasn't returned), no? 

> I admit i may be making a wrong analysis (wm4 is after all the author of this API), but what i know is that ffmpeg and ffplay all handle this sample right by calling receive_frame repeatedly until it returns EAGAIN, and making your repro program do the same also fixes the issue for it.

I don't think this is entirely correct, since threads are involved here, and the behavior is different (and what I would expect it to work like) when threads==1.

Further, I don't think we can generically wrap in a `do{}while(ret != AVERROR(EAGAIN))` since other codecs (e.g. H.264 PAFF, etc.) rely on accurately calculating frame/field reorder delay along with thread delay for frame accurate seeking (the aforementioned packet<->frame index).

Reply:
> Yes and directly below it, it says it isn't explicitly required:
> Misleading / poor doc at best if it is required.

Not needing to follow the API usage example explicitly means you can do something else than call receive_frame in a loop, like for example call send_packet in a loop, then receive_frame or whatever, as you quoted earlier. But what you should not do is ignore the return value of either function, because they tell you if things went as expected or not.

> How exactly can one VP9 superframe contain two visible frames? I thought it was always one invisible and one visible at most, per VP9 packet. And this seems to be the case when threads==1.

I don't think this sample has more than one visible frame per superframe, but avcodec_receive_frame does effectively return two different frames in two consecutive calls in at least five cases within the 500 iterations in your repro program, as both your "fix" and a do {} while(EAGAIN) loop reveal.
What you say about threading however does hint that something may be wrong somewhere in libavcodec. The two consecutive frames may not be after all generated from the same packet data, and some delay may be at play instead, because with threads == 1 i see that your repro program otherwise unmodified always returns 0 for both send_packet and receive_frame.
The generic code in decode.c handling auto inserted bsfs may not be playing nice with bsfs that effectively return more than one output packet per input packet (as it's the case with vp9_superframe_split).

> You say it does, but then describe how it doesn't (since it ditches any data from frames it hasn't returned), no?

I gave you a detailed example of how send_packet will return EAGAIN if previous packet data (buffered internally by generic code) was not pulled internally by a receive_frame call.
If you ignore that error, discard the rejected packet, and try to send another, regardless of that one being accepted or not, the dropped one is lost and decoding is ruined. This is what's happening in the repro program when threads > 1.
Try adding a bunch of printf to see what both functions return on each iteration with or without threading.

So regardless of there being a bug in libavcodec or not, you should at least check for EAGAIN in avcodec_send_packet. Consider what wm4 wrote in the relevant avcodec_send_packet call in ffmpeg.c

> // In particular, we don't expect AVERROR(EAGAIN), because we read all
   // decoded frames with avcodec_receive_frame() until done.

The latter being something you say you don't want to do.

Reply:
> Not needing to follow the API usage example explicitly means you can do something else than call receive_frame in a loop, like for example call send_packet in a loop, then receive_frame or whatever, as you quoted earlier. But what you should not do is ignore the return value of either function, because they tell you if things went as expected or not.

Certaintly not how I read it at all, and not how several others I've asked bout this have read it. I argue that the very fact that we're arguing over how to interpret the document meants it is poor / not sufficient.

> What you say about threading however does hint that something may be wrong somewhere in libavcodec. The two consecutive frames may not be after all generated from the same packet data, and some delay may be at play instead, because with threads == 1 i see that your repro program otherwise unmodified always returns 0 for both send_packet and receive_frame.
> The generic code in decode.c handling auto inserted bsfs may not be playing nice with bsfs that effectively return more than one output packet per input packet (as it's the case with vp9_superframe_split).

Curious how this one pans out.

> So regardless of there being a bug in libavcodec or not, you should at least check for EAGAIN in avcodec_send_packet.

I'll add this.

> Consider what wm4 wrote in the relevant avcodec_send_packet call in ffmpeg.c

I should not have to read `ffmpeg.c` to gain information on a public API. However, noted.

Reply:
Yeah, frame threading code delaying frames it seems. The first ten or so packets are consumed before the first frame is ever output. Try setting slice threading in your repro program, also with 16 threads. Output is fine then.

Not sure if bug in any case, seeing that actually looking at the return values of the decode API and feeding packets/pulling frames as required works around this.

Reply:
> Yeah, frame threading code delaying frames it seems. The first ten or so packets are consumed before the first frame is ever output. Try setting slice threading in your repro program, also with 16 threads. Output is fine then.

Thread delay has always been known, but it is usually constant/calculable as `avctx->thread_count -1`, but this apparently does not hold for VP9? This is kind of a pain for trying to implement generic (non-codec/format specific) frame accurate indexed seeking.

Further to that, thread delay shouldn't cause any more delay after the initial `N` packets, but that seems to not be the case here; that is, we see non-0 returns midstream we wouldn't normally see (and don't see with `threads==1`). To my knowledge, this is not correct behavior.